Add algorithm to sum banks for ISIS Reflectometry

gemmaguest opened the pull request:
**Description of work.**

This PR adds an algorithm to do the first steps of reduction from an ISIS Reflectometry 2D detector to sum across banks from a region of interest into a single bank.

This uses SmoothNeighbours to do the summing. Some minor changes were required to get this to work with INTER data:
  - Ignore slits and supermirrors in `containsRectDetectors` because they were causing the instrument to be incorrectly identified as non-rectangular.
  - Add a flag to `getDetectorIDToWorkspaceIndexMap` to ignore invalid detector IDs. This was causing an error due to having multiple detectors per spectrum, but this is irrelevant in our case where those are invalid detectors which will be ignored anyway.

**To test:**

- Download the input workspace here (note that this has had an experimental IDF applied to it).
  - [INTER45455_inst.zip](https://github.com/mantidproject/mantid/files/7453399/INTER45455_inst.zip)
- Run the script below. Note that the first two lines execute the new algorithm. The rest is a comparison with the same calculation done using GroupDetectors.
- Check that the CompareWorkspaces call at the end outputs the text `The workspaces "INTER45455_summedX" and "Inter45455_SummedX_GroupDets" matched!` in the log.

```
ws=Load('INTER45455_inst.nxs', OutputWorkspace='INTER45455')
ReflectometryISISSumBanks(InputWorkspace='INTER45455', OutputWorkspace='INTER45455_summedX', ROIDetectorIDs='2061-2180')

#
# Manual check against sum in X using GroupDetectors
#
start_idx, end_idx = 65, 184 # ws idx for our start/end det ID
npixels=ws.getInstrument().getComponentByName('linear-detector').ypixels()
groups = [""] * npixels
for ws_idx in range(ws.getNumberHistograms()):
    if ws_idx < start_idx or ws_idx > end_idx:
        continue
    # Work out which output pixel to sum the input pixel into
    bank = math.ceil(ws_idx / npixels)
    pixel = ws_idx - bank * npixels
    # Create a grouping pattern to do the sum for each output pixel
    groups[pixel] = groups[pixel] + str(ws_idx) + '+'
# Create the final grouping pattern to do all the sums
grouping = ', '.join(groups)
GroupDetectors(InputWorkspace=ws, OutputWorkspace='Inter45455_SummedX_GroupDets', GroupingPattern=grouping, Behaviour='Sum')
CompareWorkspaces(Workspace1='INTER45455_summedX', Workspace2='Inter45455_SummedX_GroupDets', CheckSpectraMap=False, CheckInstrument=False, CheckSample=True, ToleranceRelErr=True, CheckAllData=True)
```

This sums the middle two banks pixel-by-pixel into a single bank. Sliceviewer plots of the original and summed workspaces should look as follows:

Before:

![image](https://user-images.githubusercontent.com/12895056/139676374-2e010dfb-770e-4685-a255-0a49b3c47348.png)

After:

![image](https://user-images.githubusercontent.com/12895056/139676420-e301e3d2-f55a-4f59-befb-9436e3484b01.png)

Fixes #32771

Release notes to follow


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

gemmaguest committed:
Add algorithm to sum banks for ISIS Reflectometry

Re #32771

gemmaguest committed:
Ignore some components when detecting rectangular

Ignore slits and supermirrors in the Instrument method to determine if a detector is rectangular or not.

Re #32771

gemmaguest committed:
Ignore invalid detectors in map to workspace index

Re #32771

gemmaguest committed:
Ignore invalid detectors in SmoothNeighbours

This avoids a problem where getting the detector map could throw if invalid detector IDs exist in the workspace even though they will be ignored in the calculation anyway because they are not rectangular detectors.

Re #32771

gemmaguest committed:
Add error handling to ensure the expected detector type

Add checks to ensure we have a single rectangular detector

Re #32771

DavidFair commented on Framework/Geometry/src/Instrument.cpp:
```diff
@@ -1081,8 +1081,9 @@ Instrument::ContainsState Instrument::containsRectDetectors() const {
-    // skip choppers HACK!
-    if (comp->getName() == "chopper-position") {
+    // skip choppers, slits and supermirrors - HACK!
+    if (comp->getName() == "chopper-position" || comp->getName().substr(0, 4) == "slit" ||
```
Could we just get the component name once instead of 3 times?

DavidFair commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISSumBanks.py:
```diff
@@ -0,0 +1,110 @@
+    _ROI = 'ROIDetectorIDs'
+    _OUTPUT_WS = 'OutputWorkspace'
+
+    def __init__(self):
+        """Initialize an instance of the algorithm."""
+        DataProcessorAlgorithm.__init__(self)
```
This can be removed if we're not declaring any `self.foo` types

DavidFair commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISSumBanks.py:
```diff
@@ -0,0 +1,110 @@
+    def validateInputs(self):
+        issues = dict()
+        workspace = self.getProperty(self._WORKSPACE).value
+        try:
```
Do we need to check for an empty dict here?

gemmaguest replied:
I'm not sure what you mean? Isn't it supposed to return an empty dict if there aren't any errors?

DavidFair replied:
Sorry to be clear I meant if the list of ROI detectors is empty we should add an error to validate inputs
But I got the dictionary we return and the list we get in swapped when I wrote the comment

DavidFair commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISSumBanks.py:
```diff
@@ -0,0 +1,110 @@
+        result = None
+        inst = workspace.getInstrument()
+        if not inst:
+            raise RuntimeError('The input workspace must have an instrument')
```
This should be checked in validate properties since it's a fairly cheap operation (just a get) to do twice

gemmaguest replied:
This function will be called by validateInputs so it should be caught there...

DavidFair replied:
:man_facepalming: Good point :+1:

DavidFair commented on Testing/SystemTests/tests/framework/ReflectometryISISSumBanksTest.py:
```diff
@@ -0,0 +1,41 @@
+                                  OutputWorkspace=self.OUT_WS_NAME)
+
+    def validate(self):
+        return self.OUT_WS_NAME, "INTER45455_SummedX.nxs"
```
This should use `self.REF_FILENAME`

DavidFair commented on Testing/SystemTests/tests/framework/ReflectometryISISSumBanksTest.py:
```diff
@@ -0,0 +1,41 @@
+
+        # Banks 1 + 2
+        start_roi, end_roi = get_first_detid_in_bank(1),  get_first_detid_in_bank(3) - 1
+        roi_str = str(start_roi) + '-' + str(end_roi)
```
Could we use an f-string for this?
`roi_str = f"{get_first_detid_in_bank(1)}-{get_first_detid_in_bank(3) - 1}"`

DavidFair commented on Framework/API/src/MatrixWorkspace.cpp:
```diff
@@ -477,18 +477,25 @@ bool MatrixWorkspace::hasGroupedDetectors() const {
- * error
- *         if there is more than one detector for a specific workspace index.
+ * error if there is more than one detector for a specific workspace index.
+ *  @param ignoreIfNoValidDets :: set to true to exclude spectra that do not
```
If a spectra had multiple det IDs associated, some valid and some invalid, do we end up dropping all of the IDs associated with the spectra or just return the valid ones?

This might be worth clarifying in the docs string as either:
`exclude spectra (and all associated detector IDs)`
or
`exclude invalid detectors IDs from a spectra`

gemmaguest replied:
It actually returns all of them, including the invalid ones! I'm really not sure what to do in that case. I doubt it is something that exists in reality. I'll update the doc string though.

gemmaguest replied:
I'll also add a comment to the test to document the fact that this is not necessarily desired behaviour but is just not important to deal with at the moment

gemmaguest committed:
Minor tidying and commenting

- Avoid duplicate calls to getName()
- Improve comments and doc strings
- Remove unnecessary __init__ method
- Simplify python test code with use of f string and filename member variable

Re #32771

gemmaguest committed:
Add unit test to perform both mask and sum

The original sum-banks test checks that the entire workspace is summed. The new test first masks it and passes the masked workspace to be summed.

Re #32771

gemmaguest committed:
Sum all banks if ROI is not specified

Re #32771

DavidFair reviewed (dismissed):
Just tested, this works correctly with no ROI detector IDs specified too
LGTM, I'm happy to :shipit:

martyngigg commented on Framework/API/src/MatrixWorkspace.cpp:
```diff
@@ -477,18 +477,26 @@ bool MatrixWorkspace::hasGroupedDetectors() const {
+detid2index_map MatrixWorkspace::getDetectorIDToWorkspaceIndexMap(bool throwIfMultipleDets,
+                                                                  bool ignoreIfNoValidDets) const {
   detid2index_map map;
+  auto specInfo = spectrumInfo();
```
Just a quick question. Did we mean to copy the `SpectrumInfo` object here rather than just by const ref?

gemmaguest replied:
Oops, no - thanks for spotting that!

martyngigg reviewed (commented):
Just one small question about a potentially exspensive copy.

gemmaguest committed:
Avoid copying spectrum info Re #32771
